requests-for-comments/resilient-feature-flags

```diff
@@ -0,0 +1,81 @@
+# Request for comments: Resilient feature flags
+
+## Problem statement
+*Who are we building for, what are their needs (include example cases), why is this important?*
+
+We've recently realised a few limitations with feature flags when PostHog goes down. This isn't great for permanent feature flags as it causes things to break
+in unexpected ways & forces clients to think really hard about defaults & write code defensively.
+
+Ideally, our client libraries and APIs should 'just work', even if there's no error handling on the user's side.
+
+So, how do we make this happen?
+
+## Context
+
+The (simplified) way things currently work is: 
+
+Whenever a user shows up on a page, we send a request to `/decide` with their randomly generated anonymous ID to parse feature flags. Usually these people don't have
+properties attached to them, so only flags that don't depend on user properties resolve to `true` (assuming they're within the rollout bound).
+
+Things change when we call `identify()` on a user, as this updates their ID, and fires another request to `/decide` to get new flags based on not only current
+user properties, but any past properties that were set for this user.
+
+We need the database to:
+(a) Validate project tokens and find relevant team
+(b) Get feature flag definitions for the given team
+(c) Evaluate above flags using the person's properties stored in the DB
+
+Note that `decide` can run on any arbitrary anonymous ID.
+
+Caching decide responses doesn't make too much sense, since it's called usually only once per distinct ID per session. Every anonymous user has a new ID, and except
+for users coming back to the app again & again on the same day, we don't achieve resiliency as we can't cache for users never seen before, nor can we assume that properties
+haven't changed; while introducing new problems like cache invalidation.
+
+### Possible failure modes
+
+1. The database is down/unreachable: `decide` evaluation fails and returns 500
+2. The servers are down/unreachable: requests from client libraries time out / error out.
+
+(2) seems hard to defend against without a distributed app distribution (and then a resolver to go to the 'correct' app), but (1) is possible.
```

Comment:
what is a _distributed app distribution_?

Reply:
Talk about terrible wording 😅 . I basically mean multiple server deployments where one going down doesn't affect the other, like an edge server; and a load balancer than can appropriately link to healthy closest server.

Reply:
I don’t think we should build an SDK making an assumption on the underlying infrastructure. 

To build resilient software, use defensive programming: always hope for the best (uptime 100%) but always prepare for the worst (everything is down). AKA: a multi-geo deployment might help you, but you should not rely on it.

Reply:
I'm not assuming that at all, I'm just listing out what are ways things can go wrong here.

Even with defensive programming, the issue still remains that when servers go down, you stop getting flag information.

(a point below addresses this)